### PR TITLE
FOGL-8840 Add C++ support for permissions property of configuration items

### PR DESCRIPTION
**Pull request by MarkRiddoch**

### CODE REVIEW

**Review by pintomax — APPROVED** [review]

File: C/common/config_category.cpp [hunk 1]
```diff
@@ -872,6 +872,50 @@ vector<string> ConfigCategory::getOptions(const string& name) const
...
+		}
+	}
+	throw new ConfigItemNotFound();
+}
+
+/**
+ * Return true if the user has permission to update the named item
+ *
+ * @param name	The name of the configuration item to return
+ * @param username	The name of the user to test
```
Comment (ashish-jabble) [inline C/common/config_category.cpp:898]: Param should be `rolename` - The name of role user

File: docs/plugin_developers_guide/02_writing_plugins.rst [hunk 2]
```diff
@@ -636,6 +636,8 @@ Properties
      - The current value of the configuration item. This is not included when defining a set of default configuration in, for example, a plugin.
    * - properties
      - A set of items that are used in list and kvlist type items to create a list of groups of configuration items.
+   * - permissions
+     - An array of user roles that are allowed to update this configuration item. If not given then the configuration item can be updated by any user.
```
Comment (ashish-jabble) [inline docs/plugin_developers_guide/02_writing_plugins.rst:640]: and array cannot be empty

File: docs/plugin_developers_guide/02_writing_plugins.rst [hunk 3]
```diff
@@ -724,3 +726,47 @@ There is also a convenience function that can be used if you not want to define
...
+.. code-block:: JSON
+
+        'logLevel': {
+            'description': 'Minimum logging level reported for Core server',
+            'type': 'enumeration',
+            'displayName': 'Minimum Log Level',
+            'options': ['debug', 'info', 'warning', 'error', 'critical'],
+            'default': 'warning',
+            'order': '1',
+            'permissions': ['admin', 'edit']
```
Comment (ashish-jabble) [inline docs/plugin_developers_guide/02_writing_plugins.rst:765]: There is no 'edit' role name exists as such. You may add 'user' here. As we have only inbuilt roles so with 'edit' may confuse here to user.

File: docs/plugin_developers_guide/02_writing_plugins.rst [hunk 4]
```diff
@@ -724,3 +726,47 @@ There is also a convenience function that can be used if you not want to define
...
+        }
+
+In this case users with the role admin or edit are allowed to alter the configuration item.
```
Comment (ashish-jabble) [inline docs/plugin_developers_guide/02_writing_plugins.rst:768]: same as above

File: C/common/config_category.cpp [hunk 5]
```diff
@@ -872,6 +872,50 @@ vector<string> ConfigCategory::getOptions(const string& name) const
...
+ * @return bool	True if the named user can update the configuration item
+ * @throws exception if the item does not exist in the category
+ */
+bool ConfigCategory::hasPermission(const std::string& name, const std::string& username) const
```
Comment (ashish-jabble) [inline C/common/config_category.cpp:902]: For future use?

File: C/common/config_category.cpp [hunk 6]
```diff
@@ -872,6 +872,50 @@ vector<string> ConfigCategory::getOptions(const string& name) const
...
+	for (unsigned int i = 0; i < m_items.size(); i++)
+	{
+		if (name.compare(m_items[i]->m_name) == 0)
+		{
+			return m_items[i]->m_permissions;
+		}
+	}
+	throw new ConfigItemNotFound();
+}
+
```
Comment (ashish-jabble) [inline C/common/config_category.cpp:893]: As you have added getter for this, Don't we require setter for the same item?

File: docs/plugin_developers_guide/02_writing_plugins.rst [hunk 7]
```diff
@@ -724,3 +726,47 @@ There is also a convenience function that can be used if you not want to define
...
+      "admin"
+    ],
+    "value": "optional"
+  }
+
+The use of the permissions property with the single role of admin means that only admin users can change this setting. If we wished to allow more than admin users we can add another role. For example
+
+.. code-block:: JSON
+
+        'logLevel': {
```
Comment (ashish-jabble) [inline docs/plugin_developers_guide/02_writing_plugins.rst:758]: Use Double quotes in dict for the JSON highlighting.

**Review by ashish-jabble — CHANGES_REQUESTED** [review]
With the Python server changes, configuration API integration tests are failing.
We should also think of permissions KV Pair for a category, as of now only for config item. For example: For REST API category, permissions KV pair added for each config item.

File: C/common/config_category.cpp (hunk 5 above)
Comment (MarkRiddoch) [reply C/common/config_category.cpp:902]: Yes, nothing is using this currently

File: C/common/config_category.cpp (hunk 6 above)
Comment (MarkRiddoch) [reply C/common/config_category.cpp:893]: Nothing is using the currently, setting is an issue with knowing how has permissions. It is a construction time only thing to set it currently. We can tackle that if and when we need it.

**Review by pintomax — APPROVED** [review]

File: C/common/config_category.cpp (hunk 6 above)
Comment (ashish-jabble) [reply C/common/config_category.cpp:893]: Okay, I thought it was required in `setItemAttribute`. Seems like only required for an optional attribute in categoryItem constructor and its JSON representation function i.e `toJSON` and `defaultToJSON`

**Review by ashish-jabble — APPROVED** [review]

**Pull request merged**